File: leetcode/minimum-path-sum.java
public class minimum-path-sum {
    int[][] memo;
    
    public int minPathSum(int[][] grid) {
        int row = grid.length;
        int col = grid[0].length;
        
        memo = new int[row][col];
        for(int i=0; i< row; i++){
            for(int j=0; j< col; j++){
                memo[i][j] = Integer.MAX_VALUE;
            }
        }
        memo[0][0] = 0;
       
        return RecurseSum(grid, row - 1, col - 1);
        
    }

    public int RecurseSum(int[][] arr, int i, int j){
        if (i < 0 || j < 0)
            return Integer.MAX_VALUE;

        else if(i == 0 && j == 0)
            return arr[0][0];

        else
            return arr[i][j] + Math.min(
                                RecurseSum(arr, i - 1, j), 
                                RecurseSum(arr, i, j - 1));        
    }

    public static void main(){

    }
}

File: leetcode/minimum-path-sum.py
# Analysis
# Use a cache to store the distances of the minimum path sums
# Check for the position up and left for each of the fields
# -> O(m*n)
# return cache[m - 1][n - 1]



class Solution(object):
    MAX = float("inf")
    
    # Check boundaries and get cache value
    def getFieldValue(self, table, row, col):
        if row >= 0 and row < len(table) and col >=0 and col < len(table[row]):
            return table[row][col]
        return self.MAX
    
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """

        # If grid empty, don't do anything
        if not grid or not grid[0]: return 0
        

        # Help variables
        M = len(grid)
        N = len(grid[0])
        lastRow = M - 1
        lastCol = N - 1
        
        # Initialize cache for memoization, remember the best min sum
        cache = [[self.MAX] * N for i in range(M)]
        cache[0][0] = 0


        # Iterate over matrix and cache min sum
        for row in range(M):
            for col in range(N):
                cache[row][col] = grid[row][col] + min( cache[row][col], 
                                                        self.getFieldValue(cache, row - 1, col),
                                                        self.getFieldValue(cache, row, col - 1) )
        return cache[lastRow][lastCol]
                
        
        
Solution().minPathSum([[0]])